File: agent_sdks/python/a2ui_core/src/a2ui/core/schema/v1_0/agent_to_renderer.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Pydantic v2 models for Agent-to-Renderer v1.0 messages."""

from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import Field
from .common_types import CallId, FunctionCall, FunctionResponse, StrictBaseModel
from .constants import VERSION_1_0


class CreateSurface(StrictBaseModel):
    """Payload instructing the renderer to create and initialize a surface."""

    surface_id: str = Field(
        ...,
        alias="surfaceId",
        description="Globally unique identifier for the surface.",
    )
    catalog_id: Optional[str] = Field(
        None,
        alias="catalogId",
        description="Default catalog identifier for the surface.",
    )
    send_data_model: Optional[bool] = Field(
        False,
        alias="sendDataModel",
        description="Whether renderer sends full data model in action metadata.",
    )
    components: Optional[List[Dict[str, Any]]] = Field(
        None,
        description="Initial list of UI components for the surface.",
    )
    data_model: Optional[Dict[str, Any]] = Field(
        None,
        alias="dataModel",
        description="Initial root data model object.",
    )
    metadata: Optional[Dict[str, Any]] = Field(
        None,
        description="Optional surface metadata.",
    )


class CreateSurfaceMessage(StrictBaseModel):
    """Envelope wrapping a createSurface message payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    create_surface: CreateSurface = Field(
        ...,
        alias="createSurface",
        description="Create surface instruction payload.",
    )


class UpdateComponents(StrictBaseModel):
    """Payload instructing the renderer to update a surface component tree."""

    surface_id: str = Field(
        ...,
        alias="surfaceId",
        description="Identifier of the target surface.",
    )
    components: List[Dict[str, Any]] = Field(
        ...,
        description="List of updated UI components.",
    )


class UpdateComponentsMessage(StrictBaseModel):
    """Envelope wrapping an updateComponents message payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    update_components: UpdateComponents = Field(
        ...,
        alias="updateComponents",
        description="Update components payload.",
    )


# Aliases for v1.0 surface component updates
UpdateSurface = UpdateComponents
UpdateSurfaceMessage = UpdateComponentsMessage


class UpdateDataModel(StrictBaseModel):
    """Payload updating the data model of an active surface."""

    surface_id: str = Field(
        ...,
        alias="surfaceId",
        description="Identifier of target surface.",
    )
    path: Optional[str] = Field(
        None,
        description="JSON Pointer path into data model. Defaults to root path '/'.",
    )
    value: Any = Field(
        ...,
        description="Value to write or update at target path.",
    )


class UpdateDataModelMessage(StrictBaseModel):
    """Envelope wrapping an updateDataModel message payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    update_data_model: UpdateDataModel = Field(
        ...,
        alias="updateDataModel",
        description="Update data model payload.",
    )


class DeleteSurface(StrictBaseModel):
    """Payload instructing renderer to delete an active surface."""

    surface_id: str = Field(
        ...,
        alias="surfaceId",
        description="Identifier of surface to delete.",
    )


class DeleteSurfaceMessage(StrictBaseModel):
    """Envelope wrapping a deleteSurface message payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    delete_surface: DeleteSurface = Field(
        ...,
        alias="deleteSurface",
        description="Delete surface payload.",
    )


class CallRendererFunction(StrictBaseModel):
    """Payload instructing renderer to execute a function locally."""

    function_call_id: CallId = Field(
        ...,
        alias="functionCallId",
        description="Unique function call instance ID.",
    )
    call_function: FunctionCall = Field(
        ...,
        alias="callFunction",
        description="Function signature and argument payload.",
    )


class CallRendererFunctionMessage(StrictBaseModel):
    """Envelope wrapping a callRendererFunction message payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    call_renderer_function: CallRendererFunction = Field(
        ...,
        alias="callRendererFunction",
        description="Call renderer function payload.",
    )


class AgentFunctionResponseMessage(StrictBaseModel):
    """Envelope wrapping an agentFunctionResponse return payload."""

    version: Literal["v1.0"] = Field(
        VERSION_1_0, description="Protocol specification version."
    )
    agent_function_response: FunctionResponse = Field(
        ...,
        alias="agentFunctionResponse",
        description="Function response payload.",
    )


AgentToRendererMessage = Union[
    CreateSurfaceMessage,
    UpdateComponentsMessage,
    UpdateDataModelMessage,
    DeleteSurfaceMessage,
    CallRendererFunctionMessage,
    AgentFunctionResponseMessage,
]

File: typescript/web_core/src/v1_0/schema/agent-to-renderer.ts
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {z} from 'zod';
import {
  AnyComponentSchema,
  CallIdSchema,
  ExtensionsSchema,
  FunctionCallSchema,
  FunctionResponseSchema,
} from './common-types.js';

/** Schema for creating a new surface in v1.0. */
export const CreateSurfaceMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    createSurface: z
      .object({
        surfaceId: z.string().describe('The unique identifier for the UI surface to be rendered.'),
        catalogId: z
          .string()
          .optional()
          .describe('Default catalog ID for components and functions in this surface.'),
        sendDataModel: z
          .boolean()
          .optional()
          .describe('Whether the renderer sends full data model in A2A metadata.'),
        components: z
          .array(AnyComponentSchema)
          .min(1)
          .optional()
          .describe('Initial component definitions for the surface.'),
        dataModel: z
          .record(z.any())
          .optional()
          .describe('Initial root data model object for the surface.'),
        metadata: z
          .object({
            extensions: ExtensionsSchema.optional(),
          })
          .strict()
          .optional()
          .describe('Optional surface-level metadata.'),
      })
      .strict(),
  })
  .strict()
  .describe('Message sent to create a new surface container.');

/** Create surface message type. */
export type CreateSurfaceMessage = z.infer<typeof CreateSurfaceMessageSchema>;

/** Schema for updating components on an existing surface in v1.0. */
export const UpdateComponentsMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    updateComponents: z
      .object({
        surfaceId: z.string().describe('The unique identifier for the UI surface to be updated.'),
        components: z
          .array(AnyComponentSchema)
          .min(1)
          .describe('List of UI components to add or update in the surface.'),
      })
      .strict(),
  })
  .strict()
  .describe('Message sent to update component definitions on a surface.');

/** Update components message type. */
export type UpdateComponentsMessage = z.infer<typeof UpdateComponentsMessageSchema>;

/** Schema for updating the data model on a surface in v1.0. */
export const UpdateDataModelMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    updateDataModel: z
      .object({
        surfaceId: z
          .string()
          .describe('The unique identifier for the UI surface data model update.'),
        path: z
          .string()
          .optional()
          .describe('Optional path within data model. Omitted or "/" targets root.'),
        value: z.any().describe('The data value to set at the specified path.'),
      })
      .strict(),
  })
  .strict()
  .describe('Message sent to update data model values on a surface.');

/** Update data model message type. */
export type UpdateDataModelMessage = z.infer<typeof UpdateDataModelMessageSchema>;

/** Schema for deleting a surface in v1.0. */
export const DeleteSurfaceMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    deleteSurface: z
      .object({
        surfaceId: z.string().describe('The unique identifier for the UI surface to be deleted.'),
      })
      .strict(),
  })
  .strict()
  .describe('Message sent to delete a surface container.');

/** Delete surface message type. */
export type DeleteSurfaceMessage = z.infer<typeof DeleteSurfaceMessageSchema>;

/** Schema for invoking a function on the renderer in v1.0. */
export const CallRendererFunctionMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    callRendererFunction: z
      .object({
        functionCallId: CallIdSchema.describe('Unique identifier for this function call instance.'),
        callFunction: FunctionCallSchema.extend({
          catalogId: z.string().describe('Catalog ID defining the function.'),
        }).describe('Function call parameters and name.'),
      })
      .strict(),
  })
  .strict()
  .describe('Message sent to execute a function locally on the renderer.');

/** Call renderer function message type. */
export type CallRendererFunctionMessage = z.infer<typeof CallRendererFunctionMessageSchema>;

/** Schema for returning response from an agent-executed function in v1.0. */
export const AgentFunctionResponseMessageSchema = z
  .object({
    version: z.literal('v1.0'),
    agentFunctionResponse: FunctionResponseSchema.describe('Function call response payload.'),
  })
  .strict()
  .describe('Message returning the result of an agent function call.');

/** Agent function response message type. */
export type AgentFunctionResponseMessage = z.infer<typeof AgentFunctionResponseMessageSchema>;

/** Union schema for all v1.0 Agent-to-Renderer messages. */
export const AgentToRendererMessageSchema = z
  .union([
    CreateSurfaceMessageSchema,
    UpdateComponentsMessageSchema,
    UpdateDataModelMessageSchema,
    DeleteSurfaceMessageSchema,
    CallRendererFunctionMessageSchema,
    AgentFunctionResponseMessageSchema,
  ])
  .describe('Union of all v1.0 messages sent from an agent to a renderer.');

/** Agent to renderer message union type for v1.0. */
export type AgentToRendererMessage = z.infer<typeof AgentToRendererMessageSchema>;

/** Schema for a list of v1.0 agent-to-renderer messages. */
export const AgentToRendererMessageListSchema = z
  .array(AgentToRendererMessageSchema)
  .describe('Array of v1.0 agent-to-renderer messages.');

/** List of agent to renderer messages type. */
export type AgentToRendererMessageList = z.infer<typeof AgentToRendererMessageListSchema>;

/** Schema wrapping a list of v1.0 agent-to-renderer messages. */
export const AgentToRendererMessageListWrapperSchema = z
  .object({
    messages: AgentToRendererMessageListSchema,
  })
  .strict()
  .describe('Object wrapper for a list of v1.0 agent-to-renderer messages.');

/** Wrapper for list of agent to renderer messages type. */
export type AgentToRendererMessageListWrapper = z.infer<
  typeof AgentToRendererMessageListWrapperSchema
>;
